packages/opentelemetry-plugin-mongodb-core/src/mongodb.ts

```diff
@@ -0,0 +1,254 @@
+    } else {
+      return defaulType;
+    }
+  }
+
+  /**
+   * Populate span's attributes by fetching related metadata from the context
+   * @param span span to add attributes to
+   * @param ns mongodb namespace
+   * @param command mongodb internal representation of a command
+   * @param topology mongodb internal representation of the network topology
+   */
+  private _populateAttributes(
+    span: Span,
+    ns: string,
+    command: MongoInternalCommand,
+    topology: MongoInternalTopology
+  ) {
+    // add network attributes to determine the remote server
+    if (topology && topology.s && topology.s.options) {
+      span.setAttributes({
+        [AttributeNames.PEER_HOSTNAME]: `${topology.s.options.host}`,
+        [AttributeNames.PEER_PORT]: `${topology.s.options.port}`,
+      });
+    }
+    // add database related attributes
+    span.setAttributes({
+      [AttributeNames.DB_INSTANCE]: `${ns}`,
+      [AttributeNames.DB_TYPE]: `mongodb`,
+      [AttributeNames.COMPONENT]: 'mongodb-core',
+    });
+
+    if (command === undefined) return;
+    const query = Object.keys(command.query || command.q || {}).reduce(
+      (obj, key) => {
+        obj[key] = '?';
+        return obj;
+      },
+      {} as { [key: string]: string }
+    );
+    span.setAttribute('db.statement', JSON.stringify(query));
+  }
+
+  /** Creates spans for Cursor operations */
+  private _getPatchCursor() {
+    const plugin = this;
+    return (original: Func<mongodb.Cursor>) => {
+      return function patchedCursorCommand(
+        this: {
+          ns: string;
+          cmd: MongoInternalCommand;
+          topology: MongoInternalTopology;
+        },
+        ...args: unknown[]
+      ): mongodb.Cursor {
+        const currentSpan = plugin._tracer.getCurrentSpan();
+        const resultHandler = args[0] as Func<unknown> | undefined;
+        if (currentSpan === null || resultHandler === undefined) {
+          return original.apply(this, (arguments as unknown) as unknown[]);
+        }
+        const span = plugin._tracer.startSpan(`mongodb.query`, {
+          parent: currentSpan,
+          kind: SpanKind.CLIENT,
+        });
+        plugin._populateAttributes(span, this.ns, this.cmd, this.topology);
+
+        return original.call(this, plugin._patchEnd(span, resultHandler));
+      };
+    };
+  }
+
+  /**
+   * Ends a created span.
+   * @param span The created span to end.
+   * @param resultHandler A callback function.
+   */
+  private _patchEnd(span: Span, resultHandler: Func<unknown>): Function {
+    return function patchedEnd(this: {}, ...args: unknown[]) {
+      const error = args[0];
+      if (error instanceof Error) {
+        span.setStatus({
+          code: CanonicalCode.UNKNOWN,
+          message: error.message,
+        });
+      } else {
+        span.setStatus({
+          code: CanonicalCode.OK,
+        });
+      }
+      span.end();
+      return resultHandler.apply(this, (arguments as unknown) as unknown[]);
```

Comment:
What is the reason for all the type casting here?
```suggestion
      return resultHandler.apply(this, arguments);
```